src/crypto/tls/handshake_server_tls13.go

```diff
@@ -597,13 +607,26 @@ func illegalClientHelloChange(ch, ch1 *clientHelloMsg) bool {
 func (hs *serverHandshakeStateTLS13) sendServerParameters() error {
 	c := hs.c
 
+	earlySecret := hs.earlySecret
+	if earlySecret == nil {
+		earlySecret = hs.suite.extract(nil, nil)
+	}
+	hs.handshakeSecret = hs.suite.extract(hs.sharedKey,
+		hs.suite.deriveSecret(earlySecret, "derived", nil))
+
 	// Confirm ECH acceptance.
-	if hs.clientHello.encryptedClientHelloOffered &&
-		len(hs.clientHello.encryptedClientHello) == 0 {
-		secret := hs.suite.extract(hs.clientHello.random, nil)
-		context := hs.hello.random[:24]
-		acceptConfirmation := hs.suite.expandLabel(secret, echTls13LabelAcceptConfirm, context, 8)
-		copy(hs.hello.random[24:], acceptConfirmation)
+	if hs.clientHello.echIsInner {
+		confTranscript := cloneHash(hs.transcript, hs.suite.hash)
+		if confTranscript == nil {
+			c.sendAlert(alertInternalError)
+			return errors.New("tls: internal error: failed to clone hash")
+		}
+		confTranscript.Write(hs.clientHello.marshal())
```

Comment:
What do you mean by "raw representation"? If you mean: should the input to he hash begin with the ClientHello (i.e., `hs.clientHello.marshal()`), it depends on whether the HRR path has been taken. If it has, then starting with a fresh transcript would be incorrect. To prove this to yourself, replace:
```
confTranscript := cloneHash(hs.transcript, hs.suite.hash)
```
with
```
confTranscript := hs.suite.hash.New() 
```
This will cause the client to issue a "bad record MAC" alert in one of the ECH unit tests.

Reply:
Sorry, I meant should `hs.clientHello.raw` be `nil` before calling `marshal` on it, not whether or not the transcript is reset. (I agree that it shouldn't be for the reason you state.)

Reply:
(Chatted offline, since there's no CH rewriting)